### ISSUE TITLE
[QUESTION|BUG] Not Obfuscated Props

### ISSUE THREAD

**Issue by BlackYuzia** [javascript-obfuscator/javascript-obfuscator#878]
I have obfuscated code on JS:
https://pastebin.com/vLATku6x

Here is original code:
[**TS**](https://pastebin.com/q0gawKhp) | [**JS**](https://pastebin.com/y49ka6u6)

How I saw TS didn't support. But that isn't my question.

I saw what when I run code I get next output:
![image](https://user-images.githubusercontent.com/25010528/107772628-23756800-6d45-11eb-9d28-f42f4067ddda.png)
All is good, work fine. But I saw some strange result.

[**Obfuscator config**](https://pastebin.com/QivxF3g1)

How you can see in my config I was enable change (obfuscate) keys of objects, globals, properties names. So, I suggest what props like `name`, `keys`, `index`, `list` would change (obfuscated) too. But I didn't saw that result.

**So, my question-  this is a bug or not? Why some props isn't obfuscated?**

UPD:

I also a bit edit code. So I will find more "interest" things.
- Without keys obfuscate - prop `kruzya.pidoor` is exist and have value (9999).
![image](https://user-images.githubusercontent.com/25010528/107774661-f4acc100-6d47-11eb-9c9f-35e0480b8464.png)
- With keys obfuscate - prop `kruzya.pidoor` isn't exist BUT prop `kruzya._0x2bdd58`didn't have value (9999).
![image](https://user-images.githubusercontent.com/25010528/107774868-41909780-6d48-11eb-8c0a-58495c57f2c1.png)
- **With** and **without** keys obfuscate, props of class didn't obfuscate (mb because this is a class? But one prop was obfuscated. **MAGIC**)
This `"debug"` I will get by next code:
```
console.log("\n\n[KRUZYA]");
console.log("[kruzya]", kruzya);


// Auto check all exist 
console.log("\n\n[AUTO]");
for (const key in kruzya) {
    console.log(`[kruzya.${key}]`, kruzya[key])
}

// Manual check "should" exist props
console.log("\n\n[MANUAL]");
console.log("[kruzya.name]", kruzya.name);
console.log("[kruzya.pidor]", kruzya.pidor);
console.log("[kruzya.keys]", kruzya.keys);
```

**Comment by sanex3339** [comment]
Hi. Here is the list of reserved properties

**Comment by BlackYuzia** [comment]
### About props

Oh, yep.
![image](https://user-images.githubusercontent.com/25010528/107841228-ca97e500-6dc1-11eb-886e-571d1830f69e.png)
I saw what `name`, `list` and some other props would be ignored.

Did we have `anywayObfuscateNames: ["list", "name"]` setting for obfuscator?
This, probably, can destroy code, but I think if I (developer) know what I do, this shouldn't destroy all, right :eyes:?

**Note: Not sure did we should have this function. This really can destroy ALL. So, maybe we shouldn't?**

### About losed value

Wha about `undefined` ? Why I get lose value of `_0x2bdd58` prop?
`kruzya.pi...r === kruzya._0x2bdd58`, but now this value is `undefined`, not `9999`.

Can you explain me a bit more about this, pls?
I saw note about if your code is broken, do not use property changes or somethnig like that.
```
// quota
 this option WILL break your code in most cases. Enable it only if you know what it does!
```

I, also, can use `reservedNames` function and add to array my `pidor` property. But, I want to know HOW `pidor` property value was disapear.

PS: Lol, so many `blue-guys` in this issue :eyes:

**Comment by BlackYuzia** [comment]
@sanex3339 ping

I hope get answer about losed value in current moment. :eyes:

**Comment by BlackYuzia** [comment]
Okey. What I found.

1. This settings should destroy my code.
```js
    "controlFlowFlattening": true,                             // if this true
    "controlFlowFlatteningThreshold": 1,                       // and if this 1.0 (100%)
    "renameProperties": true,                                  // and if this true
                                                               // I get lost my value (get undefined)
```
2.  Enum (this is just example what done with this enum after obfuscate)
```ts
// original ts code, would be transform to js, so nvm.
enum PidorStatus {
    Clear = 0,
    Low = 1,
    Mid = 2,
    High = 3,
    Ultra = 4,
    Kruzya = 9999,
}

// obfuscated code (js)
var PidorStatus;
(function (_0x4e5341) {
    const _0x43fc34 = {
        '_0x2ae5ed': '5|0|3|2|4|1',
        '_0x2cb0d8': 'Low',
        '_0x2e2352': 'Kruzya',
        '_0x3ff165': 'High',
        '_0x61567a': 'Mid',
        '_0x1a09b8': 'Ultra',
        '_0x42bdec': 'Clear'
    };
    const _0x364709 = _0x43fc34['_0x2ae5ed']['split']('|');
    let _0x42795a = 0x0;
    while (!![]) {
        switch (_0x364709[_0x42795a++]) {
        case '0':
            _0x4e5341[_0x4e5341[_0x43fc34['_0x2cb0d8']] = 0x1] = _0x43fc34['_0x2cb0d8'];    // PidorStatus.Low === 1
            continue;
        case '1':
            _0x4e5341[_0x4e5341[_0x43fc34['_0x2e2352']] = 0x270f] = _0x43fc34['_0x2e2352']; // PidorStatus.Kruzya === 9999
            continue;
        case '2':
            _0x4e5341[_0x4e5341[_0x43fc34['_0x3ff165']] = 0x3] = _0x43fc34['_0x3ff165'];    // PidorStatus.High === 3
            continue;
        case '3':
            _0x4e5341[_0x4e5341[_0x43fc34['_0x61567a']] = 0x2] = _0x43fc34['_0x61567a'];    // PidorStatus.Mid === 2
            continue;
        case '4':
            _0x4e5341[_0x4e5341[_0x43fc34['_0x1a09b8']] = 0x4] = _0x43fc34['_0x1a09b8'];    // PidorStatus.Ultra === 4
            continue;
        case '5':
            _0x4e5341[_0x4e5341[_0x43fc34['_0x42bdec']] = 0x0] = _0x43fc34['_0x42bdec'];    // PidorStatus.Clear === 0
            continue;
        }
        break;
    }
}(PidorStatus || (PidorStatus = {})));
```
3. Call / Set field value in to `class Kruzya`.
```js
// original code
class Kruzya {
    // ...
    private _pidor: PidorStatus = PidorStatus.Kruzya      // === 9999
    // ...
    }
// obfuscated code
class Kruzya {
    // ...
   this['_0x502153'] = PidorStatus['_0x352d19'];            // === undefined
   // ...
}
```
4. What inside Enum.
```js
console['log']('[PidorStatus]', PidorStatus);
```
Return this;
![image](https://user-images.githubusercontent.com/25010528/107858879-2f455500-6e3f-11eb-8590-956167ee9128.png)
5. Initial `class Kruzya` with set prop value.
```js
// original code
const kruzya = new Kruzya(keys);
kruzya.pidor = PidorStatus.Kruzya;  // === 9999
kruzya.pidor = PidorStatus.Clear;   // === 0
kruzya.pidor = PidorStatus.Kruzya;  // === 9999
// obfuscated code
kruzya['_0x173352'] = PidorStatus['_0x352d19']; // === undefined
kruzya['_0x173352'] = PidorStatus['_0x3d45b4']; // === undefined
kruzya['_0x173352'] = PidorStatus['_0x352d19']; // === undefined
```
6. PidorStatus result
```js
// with controlFlowFlatteningThreshold 0.0
[PidorStatus] {
  '0': 'Clear',
  '1': 'Low',
  '2': 'Mid',
  '3': 'High',
  '4': 'Ultra',
  _0x5e2df5: 0,
  _0x2a17e3: 1,
  _0x5cc0ee: 2,
  _0x2cfcab: 3,
  _0x33abce: 4,
  _0xcfadd5: Infinity,
  Infinity: 'Kruzya'
}

// So, this should work now:
kruzya['_0x42b279'] = PidorStatus['_0xcfadd5']; // === Infinity
kruzya['_0x42b279'] = PidorStatus['_0x5e2df5']; // === 0
kruzya['_0x42b279'] = PidorStatus['_0xcfadd5']; // === Infinity
kruzya['_0x42b279'] = PidorStatus['_0x5e2df5'];  // === 0

// with controlFlowFlatteningThreshold 1.0
[PidorStatus] {
  '0': 'Clear',
  '1': 'Low',
  '2': 'Mid',
  '3': 'High',
  '4': 'Ultra',
  Clear: 0,
  Low: 1,
  Mid: 2,
  High: 3,
  Ultra: 4,
  Kruzya: Infinity,
  Infinity: 'Kruzya'
}

// So, 100% this should fuck up:
kruzya['_0x42b279'] = PidorStatus['_0xcfadd5']; // === undefined
kruzya['_0x42b279'] = PidorStatus['_0x5e2df5']; // === undefined
kruzya['_0x42b279'] = PidorStatus['_0xcfadd5']; // === undefined
kruzya['_0x42b279'] = PidorStatus['_0x5e2df5']; // === undefined
```

So, I suppouse, item `6.` just suck and this is a bug? Right?

### Results

How you can see, in my list (in item 6) we get fuck up with `PidorStatus["..."]` because obfuscator setup obfuscated `key-name` (`_0x5e2df5` and etc), but in `PidorStatus` this property not exist. So this will return `undefined` value. I, suppouse, this is a bug or something like that?

**Comment by BlackYuzia** [comment]
Should I rename this issue to `[BUG] Missed Value of Prop` ? :eyes:

**Comment by sanex3339** [comment]
Can you make a small but fully reproducible source code that will help reproduce this bug?

**Comment by sanex3339** [comment]
More simple example
```
var PidorStatus;
(function (PidorStatus) {
    PidorStatus[PidorStatus["Clear"] = 0] = "Clear";
    PidorStatus[PidorStatus["Low"] = 1] = "Low";
    PidorStatus[PidorStatus["Mid"] = 2] = "Mid";
    PidorStatus[PidorStatus["High"] = 3] = "High";
    PidorStatus[PidorStatus["Ultra"] = 4] = "Ultra";
    PidorStatus[PidorStatus["Kruzya"] = Infinity] = "Kruzya";
})(PidorStatus || (PidorStatus = {}));

console.log(PidorStatus.Kruzya);
console.log(PidorStatus.Clear);
```

With config:
```
module.exports = {
    compact: true,
    controlFlowFlattening: true,
    controlFlowFlatteningThreshold: 1,
    deadCodeInjection: false,
    deadCodeInjectionThreshold: 0,
    debugProtection: false,
    debugProtectionInterval: false,
    disableConsoleOutput: false,
    domainLock: [],
    exclude: [],
    forceTransformStrings: [],
    identifiersPrefix: '',
    identifiersDictionary: [],
    ignoreRequireImports: false,
    inputFileName: '',
    log: false,
    numbersToExpressions: false,
    renameGlobals: false,
    renameProperties: true,
    reservedNames: [],
    reservedStrings: [],
    rotateStringArray: false,
    seed: 0,
    selfDefending: false,
    shuffleStringArray: false,
    simplify: false,
    sourceMap: false,
    sourceMapBaseUrl: '',
    sourceMapFileName: '',
    sourceMapMode: 'separate',
    splitStrings: false,
    splitStringsChunkLength: 0,
    stringArray: false,
    stringArrayEncoding: [
        'none'
    ],
    stringArrayIndexesType: [
        'hexadecimal-number'
    ],
    stringArrayIndexShift: false,
    stringArrayWrappersChainedCalls: false,
    stringArrayWrappersCount: 0,
    stringArrayWrappersParametersMaxCount: 2,
    stringArrayWrappersType: 'variable',
    stringArrayThreshold: 0,
    target: 'browser',
    transformObjectKeys: false,
    unicodeEscapeSequence: false
};
```

**Comment by sanex3339** [comment]
Not a bug. It's how `renameProperties` option works.
In your case, you have to disable it.

With `renameProperties` option the obfuscated code looks like:
```
var PidorStatus;
(function (_0x1518d2) {
    var _0x4bf219 = {
        'xmHBg': '2|3|0|5|4|1',
        'dVPEJ': 'Mid',
        'IJpde': 'Kruzya',
        'hoXus': 'Clear',
        'fzbaC': 'Low',
        'qlBEC': 'Ultra',
        'vdbmI': 'High'
    };
    var _0x3493dc = _0x4bf219['xmHBg']['split']('|');
    var _0x38c115 = 0x0;
    while (!![]) {
        switch (_0x3493dc[_0x38c115++]) {
        case '0':
            _0x1518d2[_0x1518d2[_0x4bf219['dVPEJ']] = 0x2] = _0x4bf219['dVPEJ'];
            continue;
        case '1':
            _0x1518d2[_0x1518d2[_0x4bf219['IJpde']] = Infinity] = _0x4bf219['IJpde'];
            continue;
        case '2':
            _0x1518d2[_0x1518d2[_0x4bf219['hoXus']] = 0x0] = _0x4bf219['hoXus'];
            continue;
        case '3':
            _0x1518d2[_0x1518d2[_0x4bf219['fzbaC']] = 0x1] = _0x4bf219['fzbaC'];
            continue;
        case '4':
            _0x1518d2[_0x1518d2[_0x4bf219['qlBEC']] = 0x4] = _0x4bf219['qlBEC'];
            continue;
        case '5':
            _0x1518d2[_0x1518d2[_0x4bf219['vdbmI']] = 0x3] = _0x4bf219['vdbmI'];
            continue;
        }
        break;
    }
}(PidorStatus || (PidorStatus = {})));
console['log'](PidorStatus['Kruzya']);
console['log'](PidorStatus['Clear']);
```

`renameProperties` option works in a very simple way - it renames the property name and adds it to the cache. So for next property name with the same name value from the cache will be used. 

So in your case - properties with names `Kruzya` or `Clear` appear only once at the end of the code inside `console.log` call. So they are renamed to different names.

 String literals that were moved inside enum description to the object 
```
        'xmHBg': '2|3|0|5|4|1',
        'dVPEJ': 'Mid',
        'IJpde': 'Kruzya',
        'hoXus': 'Clear',
        'fzbaC': 'Low',
        'qlBEC': 'Ultra',
        'vdbmI': 'High'
```
, they are stored as object values, so they are not the properties that are used by `renameProperties` transformer and they are not used for renaming.

**Comment by sanex3339** [comment]
So, in your case you have to disable `renameProperties` option or `controlFlowFlattening` option

**Comment by BlackYuzia** [comment]
Okey, But if I get this in current case, I can get the same in many else cases? 
That is bad, we should do not use obfuscated function for bypass bug. That a bit weird.

**Comment by BlackYuzia** [comment]
@sanex3339 one note.

I can setup reserved names.
```js
    "reservedNames": ["Clear", "Low", "Mid", "High", "Ultra", "Kruzya"],
```
This kinda a bit fix this. But this isn't **panacea**. Because we should add every prop what would be fucked and we also should manualy check (find) props what will be destroed by `renameProperties`.

**Comment by sanex3339** [comment]
We can't fix this. Remember, if you have bugs with `renameProperties` disable it. This option is quite experimental and there is no way to improve how it works.

**Comment by BlackYuzia** [comment]
what happen?

**Comment by sanex3339** [comment]
Reopened.
I think that i can add a new option `renamePropertiesExcludeUsedStrings` (or so).
When this option is enabled all property names that also occur in other strings will be auto-excluded.


So for the source code: 

```
var PidorStatus;
(function (PidorStatus) {
    PidorStatus[PidorStatus["Clear"] = 0] = "Clear";
    PidorStatus[PidorStatus["Low"] = 1] = "Low";
    PidorStatus[PidorStatus["Mid"] = 2] = "Mid";
    PidorStatus[PidorStatus["High"] = 3] = "High";
    PidorStatus[PidorStatus["Ultra"] = 4] = "Ultra";
    PidorStatus[PidorStatus["Kruzya"] = Infinity] = "Kruzya";
})(PidorStatus || (PidorStatus = {}));

console.log(PidorStatus.Kruzya);
console.log(PidorStatus.Clear);
```

properties (`clear`, `low`, `mid` and other properties of the enum) will be excluded because they also occur as values)
```
PidorStatus[PidorStatus["Clear"] = 0] = "Clear";
                                              ^^                      ^^
                                       excluded because value here
```

**Comment by da411d** [comment]
<details>
<summary>
(offtop) sori, psihanul
</summary>
0JHQu9C40L0sINGC0Ysg0YPQttC1IDBMZlFzTkMxMExIUXNOQzcg0YHQstC+0LjQvCAi0LrRgNGD0LfRjyAwTC9RdU5DMDBMN1JnQT09Ii4g0KHRgtCw0LLRjCDQvtCx0YnQtdC/0YDQuNC90Y/RgtGL0LUgZm9vLWJhciDQuCDQvdC1IDBMTFJpOUdSMExIUmk5Q3kwTERRdWRHQjBZOD0=
</details>

**Comment by sanex3339** [comment]
I added PR with the `renamePropertiesAutoExclude` option.
https://github.com/javascript-obfuscator/javascript-obfuscator/pull/879

Please try it.

Also, looking for the better option name.

**Comment by BlackYuzia** [comment]
@sanex3339 how I can try it?
I mean, I should build package manualy?
Because I think I can install dev version by
```bash
$ npm i -D javascript-obfuscator@renamePropertiesAutoExclude # for example
``` 
and then I get all changes from #879 PR.

**Comment by sanex3339** [comment]
Will finish this PR next week.

**Comment by sanex3339** [comment]
Released as `renamePropertiesMode` option in `2.11.0` release
